lib/exclusive-lock.js

```diff
@@ -0,0 +1,157 @@
+'use strict'
+
+const Ajv = require('ajv')
+const schemas = require('./schemas/index.js')
+const jsonParseFilter = require('./json-parse-filter.js')
+
+const DEFAULT_LOCK_TTL_MS = 3000
+const DEFAULT_LOCK_REFRESH_MS = 1000
+const REQUIRED_MIN_MS_DIFF = 500
+
+const ajv = new Ajv()
+const validate = ajv.compile(schemas.instantiation_options)
+
+class ExclusiveLock {
+  #lock_name
+
+  constructor(opts) {
+    opts = opts || {}
+
+    if (!opts.log?.error) {
+      const err = new Error('A pino logger instance is required')
+      err.code = 'EINVAL'
+      throw err
+    }
+
+    this.log = opts.log
+
+    if (!validate(opts)) {
+      const err = new Error('Input validation failed')
+      err.code = 'EINVAL'
+      err.errors = validate.errors.map((e) => {
+        const {instancePath, message} = e
+        return {instancePath, message}
+      })
+      this.log.error(err)
+      throw err
+    }
+
+    const {
+      app_name
+    , log
+    , cache_connection
+    , lock_ttl_ms = DEFAULT_LOCK_TTL_MS
+    , lock_refresh_ms = DEFAULT_LOCK_REFRESH_MS
+    , lock_contents = 1 // Can be specified for more context, e.g. POD_NAME
+    , auto_refresh = true
+    } = opts
+
+    if (typeof cache_connection?.set !== 'function') {
+      const err = new TypeError('A valid cache_connection to a cache store is required')
+      err.code = 'EINVAL'
+      throw err
+    }
+    const diff = lock_ttl_ms - lock_refresh_ms
+    if (diff < REQUIRED_MIN_MS_DIFF) {
+      const err = new RangeError(
+        `lock_refresh_ms must be at least ${REQUIRED_MIN_MS_DIFF}ms less than lock_ttl_ms`
+      )
+      err.code = 'EINVAL'
+      err.meta = {
+        lock_ttl_ms
+      , lock_refresh_ms
+      , diff
+      }
+      throw err
+    }
+
+    this.cache_connection = cache_connection
+    this.app_name = app_name.replace(/[\s\W]+/g, '-')
+    this.log = log
+    this.lock_ttl_ms = lock_ttl_ms
+    this.lock_refresh_ms = lock_refresh_ms
+    this.lock_contents = lock_contents
+    this.auto_refresh = auto_refresh
+    this.#lock_name = `lock-manager:${this.app_name}`
+
+    this.got_lock = false
+    this.is_refreshing = false
+    this.refresh_timer = null
+  }
+
+  get lock_name() {
+    return this.#lock_name
+  }
+
+  async acquire() {
+    if (this.got_lock) {
+      this.log.warn('%s lock is already acquired. Must remove first.', this.lock_name)
+      return true
+    }
+    // We want this to throw if there are cache errors since that's a deal breaker
+    const got_lock = await this.cache_connection.set(
+      this.lock_name
+    , JSON.stringify(this.lock_contents)
```

Comment:
Is there any particular benefit to this being anything other than a string? we could save a stringify step by just making it a string.

I think letting this be anything else, opens the door for abuse. I'm not even sure I would allow user input on this. its kind of an implementation detail of using redis + set that you even need this.

Maybe just incrementing a counter and checking if the value is 1 or not?

Reply:
Valid points, but I did think that storing a `Date` would be a valid use-case.  Sure, it could just be a string, but it gets weird since the user could pass a Date that would get serialized on the way in, but not de-serialized on the way out.  Then, the user is passing a Date and getting back a string.

There isn't currently any use cases for anything other than the default of 1, so perhaps we keep it simple for now and force strings.